gadget/mountedfilesystem.go

```diff
@@ -186,3 +205,487 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
 		return writeFile(realSource, realTarget, preserveInDst)
 	}
 }
+
+type mountLookupFunc func(ps *PositionedStructure) (string, error)
+
+func makeStampFile(stamp string) (*osutil.AtomicFile, error) {
+	if err := os.MkdirAll(filepath.Dir(stamp), 0755); err != nil {
+		return nil, fmt.Errorf("cannot create stamp file prefix: %v", err)
+	}
+	return osutil.NewAtomicFile(stamp, 0644, 0, osutil.NoChown, osutil.NoChown)
+}
+
+func makeStamp(stamp string) error {
+	f, err := makeStampFile(stamp)
+	if err != nil {
+		return err
+	}
+	return f.Commit()
+}
+
+// MountedFilesystemUpdater assits in applying updates to a mounted filesystem
+type MountedFilesystemUpdater struct {
+	*MountedFilesystemWriter
+	backupDir   string
+	mountLookup mountLookupFunc
+}
+
+// NewMountedFilesystemUpdater returns an updater for given filesystem
+// structure, with structure content coming from provided root directory. The
+// mount is located by calling a mount lookup helper.
+func NewMountedFilesystemUpdater(rootDir string, ps *PositionedStructure, backupDir string, mountLookup mountLookupFunc) (*MountedFilesystemUpdater, error) {
+	fw, err := NewMountedFilesystemWriter(rootDir, ps)
+	if err != nil {
+		return nil, err
+	}
+	if mountLookup == nil {
+		return nil, fmt.Errorf("internal error: mount lookup helper must be provided")
+	}
+	if backupDir == "" {
+		return nil, fmt.Errorf("internal error: backup directory must not be unset")
+	}
+	fu := &MountedFilesystemUpdater{
+		MountedFilesystemWriter: fw,
+		backupDir:               backupDir,
+		mountLookup:             mountLookup,
+	}
+	return fu, nil
+}
+
+func fsStructBackupPath(backupDir string, ps *PositionedStructure) string {
+	return filepath.Join(backupDir, fmt.Sprintf("struct-%v", ps.Index))
+}
+
+// entryNames resolves the source, destination and backup locations for given
+// source/target combination. The source location is always within the root
+// directory passed during initialization. Backup location is within provided
+// backup directory or empty if directory was not provided.
+func (f *MountedFilesystemUpdater) entryNames(dstRoot, source, target, backupDir string) (srcName, dstName, backupName string) {
+	srcName = f.entrySource(source)
```

Comment:
srcName is then just returned.  Wondering if should have instead to independent methods:

* entrySource
* entryDests or entryTargets